src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferReadOnlySpanPropertiesOverReadOnlyArrayFields.cs

```diff
@@ -0,0 +1,493 @@
+                    IObjectCreationOperation objectCreation => objectCreation.Constructor,
+                    _ => null
+                };
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// Visits the parents of <see cref="IArrayElementReferenceOperation"/>s and eliminates candidates
+        /// who's elements are used in ways that prohibit conversion to <see cref="ReadOnlySpan{T}"/>.
+        /// </summary>
+        private sealed class ArrayElementReferenceVisitor : OperationVisitor<VisitContext, Unit>
+        {
+            private readonly Cache _cache;
+
+            public ArrayElementReferenceVisitor(Cache cache)
+            {
+                _cache = cache;
+            }
+
+            public override Unit VisitSimpleAssignment(ISimpleAssignmentOperation operation, VisitContext argument)
+            {
+                if (operation.Target.Equals(argument.Operation))
+                    _cache.RemoveCandidate(argument.Field);
+                return base.VisitSimpleAssignment(operation, argument);
+            }
+
+            public override Unit VisitCompoundAssignment(ICompoundAssignmentOperation operation, VisitContext argument)
+            {
+                if (operation.Target.Equals(argument.Operation))
+                    _cache.RemoveCandidate(argument.Field);
+                return base.VisitCompoundAssignment(operation, argument);
+            }
+
+            public override Unit VisitTuple(ITupleOperation operation, VisitContext argument)
+            {
+                if (operation.Parent is IDeconstructionAssignmentOperation deconstruction && deconstruction.Target.Equals(operation))
+                {
+                    _cache.RemoveCandidate(argument.Field);
+                }
+                return base.VisitTuple(operation, argument);
+            }
+
+            public override Unit VisitIncrementOrDecrement(IIncrementOrDecrementOperation operation, VisitContext argument)
+            {
+                _cache.RemoveCandidate(argument.Field);
+                return base.VisitIncrementOrDecrement(operation, argument);
+            }
+        }
+
+        private sealed class Cache : IDisposable
+        {
+            private readonly ImmutableHashSet<ITypeSymbol> _supportedArrayElementTypes;
+            private readonly ImmutableHashSet<IMethodSymbol> _asSpanMethods;
+            private readonly PooledConcurrentSet<(IFieldSymbol Field, IOperation Operation)> _savedOperations;
+            private readonly PooledConcurrentSet<IFieldSymbol> _candidates;
+
+            //  We keep track of whether a field symbol has a valid initializer by adding it to this collection
+            //  when we analyze its valid field reference. This is done to allow us to discard fields
+            //  that don't have any initializer.
+            private readonly PooledConcurrentSet<IFieldSymbol> _fieldsWithValidFieldInitializer;
```

Comment:
After removing the Symbol analysis for fields hopefully we will not need to use 3 caches